File: dsharp-rdp/Tokenizer.js
class Tokenizer {
    init(string) {
        this._string = string;
        this._cursor = 0;
    }

    hasMoreTokens() {
        return this._cursor < this._string.length;
    }

    getNextToken() {
        if(!this.hasMoreTokens()) {
            return null;
        }

        const string = this._string.slice(this._cursor);

        let matched = /^\d+/.exec(string);
        if(matched !== null) {
            this._cursor += matched[0].length;
            return {
                type: 'NUMBER',
                value: matched[0],
            };
        }

        matched = /^"[^"]*"/.exec(string);
        if(matched !== null) {
            this._cursor += matched[0].length;
            return {
                type: 'STRING',
                value: matched[0],
            };
        }

        matched = /^'[^']*'/.exec(string);
        if(matched !== null) {
            this._cursor += matched[0].length;
            return {
                type: 'STRING',
                value: matched[0],
            };
        }
        // Numbers:

        // if(!Number.isNaN(Number(string[0]))) {
        //     let number = '';
        //     while(this._cursor < this._string.length && !Number.isNaN(Number(string[this._cursor]))) {
        //         number += string[this._cursor++];
        //     }
        //     return {
        //         type: 'NUMBER',
        //         value: number,
        //     };
        // }

        // // String:
        // if(string[0] === '"') {
        //     let s = '';
        //     do {
        //         s += string[this._cursor++];
        //     } while(string[this._cursor]!== '"' && !this.isEOF());
        //     s += string[this._cursor++]; // skip "
        //     return {
        //         type: 'STRING',
        //         value: s,
        //     };
        // }

        return null;
    }

    isEOF() {
        return this._cursor === this._string.length;
    }
}

module.exports = {
    Tokenizer,
}

File: cs/parser/Tokenizer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace DanBuilds
{
    public class Tokenizer
    {
        private string _buffer;
        private int _cursor;

        public void Init(string buffer)
        {
            this._buffer = buffer;
            this._cursor = 0;
        }

        public bool HasMoreToken()
        {
            return this._cursor < this._buffer.Length;
        }

        public string Match(string pattern, string input)
        {
            var matched = Regex.Match(input, pattern);
            if (!matched.Success)
                return null;
            this._cursor += matched.Value.Length;
            return matched.Value;
        }
        
        public Ast<T> GetNextToken<T>()
        {
            if(!this.HasMoreToken())
            {
                return null;
            }

            var buffer = this._buffer.Slice(this._cursor);
            foreach(var spec in Specs)
            {
                var pattern = spec.Pattern;
                var tokenType = spec.TokenType;

            }
        }
    }
}
